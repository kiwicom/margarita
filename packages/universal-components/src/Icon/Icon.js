// @flow

import * as React from 'react';
import { Text as RNText } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import { Text } from '../Text';
import {
  StyleSheet,
  type PlatformStyleObjectType,
} from '../PlatformStyleSheet';
import iconsMap from './icons.json';
import type { Props } from './IconTypes';

const sanitizeIconCharacter = (iconCharacter: string) => {
  if (/^E(.{3})$/.test(iconCharacter)) {
    return String.fromCharCode(parseInt(iconCharacter, 16));
  }
  return iconCharacter;
};

export default function Icon({
  name,
  color = '#46515e',
  size = 'medium',
  style,
}: Props) {
  if (name == null) {
    return <Text style={[styles[size], { color }, style]}>?</Text>;
  }

  const icon = iconsMap[name];
  if (!icon) {
    return <Text style={[styles[size], { color }, style]}>?</Text>;
  }

  return (
    <RNText style={[styles.icon, styles[size], { color }, style]}>
      {sanitizeIconCharacter(icon.character)}
    </RNText>
  );
}

const getSizeStyle = (size: number): $Values<PlatformStyleObjectType> => ({
  fontSize: size,
  width: size,
  height: size,
  lineHeight: size,
});

const styles = StyleSheet.create({
  icon: {
    fontFamily: 'orbit-icons',
    android: {
      includeFontPadding: false,
      textAlignVertical: 'center',
    },
  },
  /* eslint-disable react-native/no-unused-styles */
  small: getSizeStyle(parseFloat(defaultTokens.widthIconSmall)),
  medium: getSizeStyle(parseFloat(defaultTokens.widthIconMedium)),
  large: getSizeStyle(parseFloat(defaultTokens.widthIconLarge)),
  /* eslint-enable */
});
