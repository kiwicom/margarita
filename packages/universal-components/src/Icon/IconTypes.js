// @flow

import type { StylePropType } from '../PlatformStyleSheet/StyleTypes';
import type { IconNameType } from '../types/_generated-types';

type Size = 'small' | 'medium' | 'large';

export type Props = {|
  +name?: IconNameType,
  size?: Size,
  color?: string,
  style?: StylePropType,
|};
