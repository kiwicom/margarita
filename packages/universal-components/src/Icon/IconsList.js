// @flow strict

import * as React from 'react';
import { FlatList, View } from 'react-native';

import { Text } from '../Text';
import { StyleSheet } from '../PlatformStyleSheet';
import iconsMap from './icons.json';

import { Icon } from '.';

const keyExtractor = item => item;
const renderItem = ({ item }) => (
  <View key={item} style={styles.item}>
    <Text>{item}</Text>
    <Icon name={item} size="large" />
  </View>
);

export default function IconsList() {
  return (
    <FlatList
      data={Object.keys(iconsMap)}
      keyExtractor={keyExtractor}
      showsVerticalScrollIndicator={false}
      contentContainerStyle={styles.list}
      numColumns={2}
      renderItem={renderItem}
    />
  );
}

const styles = StyleSheet.create({
  list: {
    paddingVertical: 10,
    paddingHorizontal: 10,
  },
  item: {
    flex: 1,
    alignItems: 'center',
    paddingHorizontal: 20,
    paddingVertical: 10,
  },
});
