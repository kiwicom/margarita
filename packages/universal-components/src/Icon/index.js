// @flow

export { default as Icon } from './Icon';
