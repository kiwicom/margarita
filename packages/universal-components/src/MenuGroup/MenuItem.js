// @flow

import * as React from 'react';
import { View, Platform } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import Separator from './Separator';
import { Icon } from '../Icon';
import { StyleSheet, type StylePropType } from '../PlatformStyleSheet';
import { Text } from '../Text';
import { Touchable } from '../Touchable';

type Props = {|
  +title: React.Node,
  +onPress: () => void,

  +titleStyle?: StylePropType,
  +icon?: React.Element<typeof Icon>,
  +subTitle?: React.Node,
  +subTitleStyle?: StylePropType,
  +label?: React.Node,
  +labelStyle?: StylePropType,
  +hideActionIcon?: boolean,
|};

export default function MenuItem({
  title,
  onPress,
  icon,
  titleStyle,
  subTitle,
  subTitleStyle,
  label,
  labelStyle,
  hideActionIcon = false,
}: Props) {
  const menuIcon =
    icon != null
      ? React.cloneElement(icon, {
          color: icon.props.color ?? defaultTokens.paletteInkLight,
          size: icon.props.size ?? 'medium',
          style: [styles.icon, icon.props.style],
        })
      : null;
  return (
    <View style={styles.container}>
      <Touchable onPress={onPress} style={styles.wrapper}>
        <View style={styles.row}>
          {menuIcon}
          <View>
            {label != null && (
              <Text type="secondary" size="small" style={labelStyle}>
                {label}
              </Text>
            )}
            <Text type="attention" size="large" style={titleStyle}>
              {title}
            </Text>
            {subTitle != null && (
              <Text type="secondary" size="small" style={subTitleStyle}>
                {subTitle}
              </Text>
            )}
          </View>
          {Platform.OS === 'ios' && hideActionIcon === false && (
            <View style={styles.actionIcon}>
              <Icon
                name="chevron-right"
                color={defaultTokens.paletteProductNormal}
              />
            </View>
          )}
        </View>
      </Touchable>
      <Separator style={icon == null ? styles.noIcon : styles.iconSeparator} />
    </View>
  );
}

const wrapperPaddingStart = 12;
const iconWidth = 24;
const iconMargin = 12;

const styles = StyleSheet.create({
  container: {
    marginTop: StyleSheet.hairlineWidth,
    ios: {
      marginBottom: -StyleSheet.hairlineWidth,
    },
  },
  wrapper: {
    backgroundColor: defaultTokens.paletteWhite,
    ios: {
      paddingTop: 8.5,
      paddingBottom: 11.5,
    },
    paddingTop: 6,
    paddingBottom: 8.5,

    paddingStart: wrapperPaddingStart,
    paddingEnd: 16,
  },
  noIcon: {
    ios: {
      marginStart: 12,
    },
  },
  iconSeparator: {
    ios: {
      marginStart: wrapperPaddingStart + iconWidth + iconMargin,
    },
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  icon: {
    marginEnd: iconMargin,
  },
  actionIcon: {
    flex: 1,
    alignItems: 'flex-end',
  },
});
