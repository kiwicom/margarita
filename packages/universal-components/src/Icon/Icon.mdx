---
name: Icon
menu: Components
route: /icon
---

import { Playground, PropsTable } from 'docz'
import { View } from 'react-native'
import { Icon } from '.'
import IconsList from './IconsList'


# Icon

TODO

## Properties

<PropsTable of={Icon} />

## Basic usage

<Playground>
  <Icon name="check" />
</Playground>

## Icons List

<IconsList />

