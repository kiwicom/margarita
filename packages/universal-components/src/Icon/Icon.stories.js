// @flow

import * as React from 'react';
import { storiesOf } from '@storybook/react-native';

import IconsList from './IconsList';

import { Icon } from '.';

storiesOf('Icon', module)
  .addDecorator(getStory => getStory())
  .add('Regular Icon', () => (
    <React.Fragment>
      <Icon name="check" />
    </React.Fragment>
  ))
  .add('Custom Icon', () => (
    <React.Fragment>
      <Icon name="check" color="#46B655" size="large" />
    </React.Fragment>
  ))
  .add('Wrong Icon Name', () => (
    <React.Fragment>
      {/* $FlowExpectedError: Wrong Icon name for showcase purposes */}
      <Icon name="-" color="#46B655" size="large" />
    </React.Fragment>
  ))
  .add('All Icons list', () => (
    <React.Fragment>
      <IconsList />
    </React.Fragment>
  ));
