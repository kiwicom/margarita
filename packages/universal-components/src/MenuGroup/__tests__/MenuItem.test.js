// @flow

import * as React from 'react';
import { shallow } from 'react-native-testing-library';

import MenuItem from '../MenuItem';
import { Icon } from '../../Icon';

describe('MenuGroup', () => {
  it('renders with subTitle', () => {
    const { output } = shallow(
      <MenuItem onPress={jest.fn()} title="test" subTitle="sub" />,
    );
    expect(output).toMatchSnapshot();
  });

  it('renders with label', () => {
    const { output } = shallow(
      <MenuItem onPress={jest.fn()} title="test" label="label" />,
    );
    expect(output).toMatchSnapshot();
  });

  it('renders with icon', () => {
    const { output } = shallow(
      <MenuItem
        onPress={jest.fn()}
        title="test"
        icon={<Icon name="clock" />}
      />,
    );
    expect(output).toMatchSnapshot();
  });
});
