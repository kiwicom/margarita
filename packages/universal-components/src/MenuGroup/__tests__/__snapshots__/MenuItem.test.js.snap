// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuGroup renders with icon 1`] = `
<Component
  style={
    Object {
      "marginBottom": -0.5,
      "marginTop": 0.5,
    }
  }
>
  <Touchable
    accessibilityComponentType="button"
    accessibilityTraits="button"
    borderlessRipple={false}
    disabled={false}
    onPress={[MockFunction]}
    rippleColor="rgba(0, 0, 0, 0.32)"
    style={
      Object {
        "backgroundColor": "#fff",
        "paddingBottom": 11.5,
        "paddingEnd": 16,
        "paddingStart": 12,
        "paddingTop": 8.5,
      }
    }
  >
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Icon
        color="#7f91a8"
        name="clock"
        size="medium"
        style={
          Array [
            Object {
              "marginEnd": 12,
            },
            undefined,
          ]
        }
      />
      <Component>
        <Text
          expo={false}
          size="large"
          type="attention"
        >
          test
        </Text>
      </Component>
      <Component
        style={
          Object {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      >
        <Icon
          color="#00a991"
          name="chevron-right"
        />
      </Component>
    </Component>
  </Touchable>
  <Separator
    style={
      Object {
        "marginStart": 48,
      }
    }
  />
</Component>
`;

exports[`MenuGroup renders with label 1`] = `
<Component
  style={
    Object {
      "marginBottom": -0.5,
      "marginTop": 0.5,
    }
  }
>
  <Touchable
    accessibilityComponentType="button"
    accessibilityTraits="button"
    borderlessRipple={false}
    disabled={false}
    onPress={[MockFunction]}
    rippleColor="rgba(0, 0, 0, 0.32)"
    style={
      Object {
        "backgroundColor": "#fff",
        "paddingBottom": 11.5,
        "paddingEnd": 16,
        "paddingStart": 12,
        "paddingTop": 8.5,
      }
    }
  >
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Component>
        <Text
          expo={false}
          size="small"
          type="secondary"
        >
          label
        </Text>
        <Text
          expo={false}
          size="large"
          type="attention"
        >
          test
        </Text>
      </Component>
      <Component
        style={
          Object {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      >
        <Icon
          color="#00a991"
          name="chevron-right"
        />
      </Component>
    </Component>
  </Touchable>
  <Separator
    style={
      Object {
        "marginStart": 12,
      }
    }
  />
</Component>
`;

exports[`MenuGroup renders with subTitle 1`] = `
<Component
  style={
    Object {
      "marginBottom": -0.5,
      "marginTop": 0.5,
    }
  }
>
  <Touchable
    accessibilityComponentType="button"
    accessibilityTraits="button"
    borderlessRipple={false}
    disabled={false}
    onPress={[MockFunction]}
    rippleColor="rgba(0, 0, 0, 0.32)"
    style={
      Object {
        "backgroundColor": "#fff",
        "paddingBottom": 11.5,
        "paddingEnd": 16,
        "paddingStart": 12,
        "paddingTop": 8.5,
      }
    }
  >
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Component>
        <Text
          expo={false}
          size="large"
          type="attention"
        >
          test
        </Text>
        <Text
          expo={false}
          size="small"
          type="secondary"
        >
          sub
        </Text>
      </Component>
      <Component
        style={
          Object {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      >
        <Icon
          color="#00a991"
          name="chevron-right"
        />
      </Component>
    </Component>
  </Touchable>
  <Separator
    style={
      Object {
        "marginStart": 12,
      }
    }
  />
</Component>
`;
